File: leetcode/2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        unordered_set<long long> bag;
        long long res = 0, countingSum = 0;
        for (int i = 0, j = 0; i < nums.size(); i++) {
            auto curr = nums[i];
            while (bag.count(curr) || bag.size() == k) {
                auto toErase = nums[j++];
                bag.erase(toErase), countingSum -= toErase;
            }
            bag.insert(curr), countingSum += curr;
            if (bag.size() == k) res = max(res, countingSum);
        }
        return res;
    }
};

File: leetcode/2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.rs
impl Solution {
    pub fn maximum_subarray_sum(nums: Vec<i32>, k: i32) -> i64 {
        let mut bag = std::collections::HashSet::new();
        let (mut res, mut j) = (0, 0);
        let mut counting_sum: i64 = 0;
        for i in 0..nums.len() {
            let curr = nums[i];
            while bag.contains(&curr) || bag.len() == k as usize {
                let to_erase = nums[j];
                j += 1;
                bag.remove(&to_erase);
                counting_sum -= to_erase as i64;
            }
            bag.insert(curr);
            counting_sum += curr as i64;
            if bag.len() == k as usize {
                res = std::cmp::max(res, counting_sum);
            }
        }
        res as i64
    }
}